File: conformance/tests/basic-p2pkh/P2PKH.runar.zig
const runar = @import("runar");

pub const P2PKH = struct {
    pub const Contract = runar.SmartContract;

    pkh: runar.Addr,

    pub fn init(pkh: runar.Addr) P2PKH {
        return .{ .pkh = pkh };
    }

    pub fn unlock(self: *const P2PKH, sig: runar.Sig, pubKey: runar.PubKey) void {
        runar.assert(runar.hash160(pubKey) == self.pkh);
        runar.assert(runar.checkSig(sig, pubKey));
    }
};

File: conformance/tests/java-p2pkh/P2PKH.runar.java
package runar.examples.p2pkh;

import runar.lang.*;

public class P2PKH extends SmartContract {
    @Readonly Addr pubKeyHash;

    public P2PKH(Addr pubKeyHash) {
        super(pubKeyHash);
        this.pubKeyHash = pubKeyHash;
    }

    @Public
    public void unlock(Sig sig, PubKey pubKey) {
        assertThat(hash160(pubKey).equals(pubKeyHash));
        assertThat(checkSig(sig, pubKey));
    }
}
